src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt

```diff
@@ -56,107 +63,164 @@ interface CommandLineProgram {
+  private val restore: () -> Unit
+) : Closeable {
+  companion object {
+    fun open(): WorkerIO {
+      val stdErr = System.err
+      val stdIn = BufferedInputStream(System.`in`)
+      val stdOut = System.out
+      val inputBuffer = ByteArrayInputStream(ByteArray(0))
+      val execution = ByteArrayOutputStream()
+      val outputBuffer = PrintStream(execution)
+
+      // delegate the system defaults to capture execution information
+      System.setErr(outputBuffer)
+      System.setOut(outputBuffer)
+      System.setIn(inputBuffer)
+
+      return WorkerIO(stdIn, stdOut, execution) {
+        System.setOut(stdOut)
+        System.setIn(stdIn)
+        System.setErr(stdErr)
       }
     }
-    return args
+
+    fun noop(): WorkerIO {
+      val inputBuffer = ByteArrayInputStream(ByteArray(0))
+      val execution = ByteArrayOutputStream()
+      val outputBuffer = PrintStream(execution)
+      return WorkerIO(inputBuffer, outputBuffer, execution) {}
+    }
+  }
+
+  override fun close() {
+    restore.invoke()
   }
 }
 
-class InvocationWorker(
-  private val delegate: CommandLineProgram,
-  private val buffer: ByteArrayOutputStream
-) {
+/** PersistentWorker follows the Bazel worker protocol and executes a CommandLineProgram. */
+class PersistentWorker(
+  private val io: WorkerIO,
+  private val program: CommandLineProgram
+) : Worker {
+  val logger = Logger.getLogger(PersistentWorker::class.java.canonicalName)
+
+  enum class Status {
+    OK, INTERRUPTED, ERROR
+  }
+
+  override fun run(args: List<String>): Int {
+    while (true) {
+      val request = WorkRequest.parseDelimitedFrom(io.input) ?: continue
 
-  fun invoke(args: List<String>): Pair<Int, String> = WorkingDirectoryContext.newContext()
-      .use { wdCtx ->
-        return try {
-          delegate.apply(wdCtx.dir, args)
-        } catch (e: RuntimeException) {
-          if (e.wasInterrupted()) BazelWorker.INTERRUPTED_STATUS
-          else BazelWorker.ERROR_STATUS.also {
-            System.err.println(
-                "ERROR: Worker threw uncaught exception with args: ${args}"
-            )
-            e.printStackTrace(System.err)
+      val (status, exit) = WorkingDirectoryContext.newContext()
+        .runCatching {
+          request.argumentsList
+            ?.let {
+              maybeExpand(it)
+            }
+            .run {
+              Status.OK to program.apply(dir, maybeExpand(request.argumentsList))
+            }
+        }
+        .recover { e: Throwable ->
+          io.execution.write((e.message ?: e.toString()).toByteArray(UTF_8))
+          if (!e.wasInterrupted()) {
+            logger.log(Level.SEVERE,
+                       "ERROR: Worker threw uncaught exception",
+                       e)
+            Status.ERROR to 1
+          } else {
+            Status.INTERRUPTED to 1
           }
-        } to buffer.toString()
+        }
+        .getOrThrow()
+
+      val response = with(WorkResponse.newBuilder()) {
```

Comment:
Nit: With is fine, but also
```
val response = WorkResponse.newBuilder().apply {
  output = String(io.execution.toByteArray(), UTF_8)
  exitCode = exit
  setRequestId(request.requestId)
}.build()
```

It's a more common pattern for working with java builders, though it's the same thing.

Reply:
Ah, cargo-culted the old code. Much cleaner, as with requires using the builder function on the last line.